File: python/src/patterns/abstractfactory/concrete_factory2.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .product_a2 import ProductA2
from .product_b2 import ProductB2

if TYPE_CHECKING:
    from .abstract_product_a import AbstractProductA
    from .abstract_product_b import AbstractProductB


class ConcreteFactory2:
    """Creates the products of family 2."""

    def create_product_a(self) -> AbstractProductA:
        return ProductA2()

    def create_product_b(self) -> AbstractProductB:
        return ProductB2()

File: typescript/src/abstract-factory/concrete-factory2.ts
import type { AbstractFactory } from "./abstract-factory.ts";
import type { AbstractProductA } from "./abstract-product-a.ts";
import type { AbstractProductB } from "./abstract-product-b.ts";
import { ProductA2 } from "./product-a2.ts";
import { ProductB2 } from "./product-b2.ts";

/** Creates the products of family 2. */
export class ConcreteFactory2 implements AbstractFactory {
  createProductA(): AbstractProductA {
    return new ProductA2();
  }

  createProductB(): AbstractProductB {
    return new ProductB2();
  }
}
